File: src/task_3_1/CarSimulator.java
package task_3_1;

public class CarSimulator {
    public CarBuilder carBuilder;

    public Car car;

    public CarSimulator(CarBuilder carBuilder, Car car) {
        this.carBuilder = carBuilder;
        this.car = car;
    }

    public void showProcess() {
        String result = String.format("%s build a car with %s wheel and %s engine:)", this.carBuilder.name, this.car.wheel.name, this.car.engine.name);
        System.out.println(result);
    }

}

File: src/task_2_2_flyweight/CarSimulator.kt
package task_2_2_flyweight

import java.util.*

class CarSimulator {
    private val cars = mutableListOf<Car>()
    private val random = Random()

    fun createCar(wheel: Wheel, engine: Engine, color: CarColor) {
        val type = CarBuilder.getCarType(wheel, engine, color)
        val number = random.nextInt(1000).toString()
        cars.add(Car(number, type))
    }

    fun showCars() {
        cars.forEach {
            it.show()
        }
    }

}